### ISSUE TITLE
GetGrain with grainClassNamePrefix not working

### ISSUE THREAD

**Issue by ZeBobo5** [dotnet/orleans#6765]
Hi,

I implemented IMyGrainInterface in 2 classes in separed silo host. 
But when I call this in silo client, only one of this two method works but not both. Why ? :
```csharp
client.GetGrain<IMyGrainInterface>(Guid.NewGuid(), "A.B.MyGrainImplementation");
client.GetGrain<IMyGrainInterface>(Guid.NewGuid(), "A.C.MyGrainImplementation");
```
The implementation classes have same name.
Thanks for help

**Comment by ReubenBond** [comment]
Which works and what is the error message?
What are the full names of the grain implementation classes?

**Comment by ZeBobo5** [comment]
I checked that all silo are up before calling client code.

The error message is 
`Cannot find an implementation class for grain interface: IMyGrainInterface, grainClassNamePrefix: A.B.MyGrainInterface. Make sure the grain assembly was correctly deployed and loaded in the silo.`
But it depends. it can be for second implementation. Thinking about the order of silo starting.

FullNames are :
- Resa.NerveCenter.Core.Services.RemoteConfigurationService
- Resa.NerveCenter.IdentityServer.Services.RemoteConfigurationService

**Comment by benjaminpetit** [comment]
Can you show how you configure the silos and the client?

**Comment by ZeBobo5** [comment]
For my client :
![Capture d’écran 2020-10-01 181027](https://user-images.githubusercontent.com/9378633/94835026-6aeffd00-0411-11eb-9909-794bf0a7880e.png)![Capture d’écran 2020-10-01 180850](https://user-images.githubusercontent.com/9378633/94834900-43009980-0411-11eb-8f5e-824f1457dd7d.png)

For silos, it's more difficult due to dependency injections, but this is the base builder for the 2 silos :
![Capture d’écran 2020-10-01 181212](https://user-images.githubusercontent.com/9378633/94835206-aee30200-0411-11eb-9c3f-6284a85ed9dc.png)

**Comment by ZeBobo5** [comment]
To be more explicit :
- Silo 1:
-- A.**B**.MyGrainImplementation
-Silo 2:
-- A.**C**.MyGrainImplementation

**Comment by benjaminpetit** [comment]
To be sure that I uderstood: A.**B**.MyGrainImplementation only exists in Silo 1, and A.**C**.MyGrainImplementation only exists in SIlo 2, correct?

**Comment by ZeBobo5** [comment]
This is correct. Grains are only implemented the same grain interface.

**Comment by ZeBobo5** [comment]
@ReubenBond , @benjaminpetit 
On this method : https://github.com/dotnet/orleans/blob/dc76212f1cc2df08f5c69cbfb02cdbae472cc89e/src/Orleans.Core/Runtime/GrainInterfaceMap.cs#L60
Implementations of map.typeToInterfaceData and map.table... have not been added if ContainsKey is true. I think this where is my problem

**Comment by benjaminpetit** [comment]
Yes, that's what I was suspecting, an issue in the `GrainInterfaceMap` merge. I assume we need to merge the `GrainInterfaceData` if we have two different values for the same grain interface.

**Comment by ZeBobo5** [comment]
It could be interesting to have a IGrainPrefixFactory, based on namespace for default implementation

**Comment by ReubenBond** [comment]
This ought to be fixed in 7.0, since the way that cluster manifests are merged has been rewritten and I believe it does not have this issue.
